content/blog/pairisto/2023-09-10_[gsoc_2023]_vpn_support_project_final_report.md

```diff
@@ -0,0 +1,42 @@
+4. Add a separate function for ethernet bridging for the driver.
+
+## What Have I Learned
+This summer taught me way more about programming concepts that I never really encountered before like mutexes, semaphores, condition variables,  networking, driver development, and even how to make my workflow more efficient and so much more than I thought it was ever going to teach me.
+
+Kernel and system software development have been central to my project as it involved delving deep into thread safety mechanisms, how to write implementations of the common driver I/O functions for a driver, and understanding the intricacies of how an operating system's network stack functions and subsequently designing an interface to complement it.
+
+On the debugging front, I acquired significant insights into addressing kernel panics and anomalies in the driver and interface behavior. This was achieved through innovative use of tools like `ghidra` and methods such as inserting debugging statements, especially given the absence of a robust debugger in our environment.
+
+Another pivotal aspect of my experience was navigating and analyzing undocumented code, primarily from OpenVPN for application testing and from Haiku for driver development. This analytical process not only aided in problem-solving but also inspired me to pen down seven blog posts, capturing and documenting the solutions I discerned.
+
+## Code Submitted
+My main TUN/TAP driver/interface commit has yet to be pushed upstream due to waiting on feedback from one of my reviewers but from what I can tell it is mostly ready to be pushed. the commits that I have made are:
+1. [Main Driver/Interface Commit](https://review.haiku-os.org/c/haiku/+/6608/)
+    1. [Indirect Ancestor to add to build definitions](https://review.haiku-os.org/c/haiku/+/6898/2?usp=related-change)
+2. [WIP Documentation for the Driver and Interface](https://review.haiku-os.org/c/haiku/+/6904/1?usp=related-change)
+    1. Here is where you can find the documentation for the driver along with how to use it via OpenVPN though it is still in a heavy work in progress
+
+## Usage
+While you can't use my code now, whenever it does get pushed in the future then one of the best ways that I have found to test it is using OpenVPN. You can download and use the modified version of OpenVPN needed to run on Haiku via [haikuporter/haikuports](https://github.com/haikuports/haikuporter).  After downloading it and installing it, you can navigate to this link [here](https://openvpn.net/community-resources/how-to/#setting-up-your-own-certificate-authority-ca-and-generating-certificates-and-keys-for-an-openvpn-server-and-multiple-clients) from OpenVPN to help with setting up the environment. You can do either a Haiku to Haiku connection or Haiku to Linux connection, all up to you.
```

Comment:
OpenVPN packages are available, no need to build them.

Reply:
Changed that in the newest update